Improve getting started experience

If I generate an app from start.spring.io with version `2.1.2.RELEASE`, I can build it and run it. However if I upgrade to `3.0.0.RELEASE` I get the following:

```
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'beanCatalogService' defined in class path resource [org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfiguration.class]: Unsatisfied dependency expressed through method 'beanCatalogService' parameter 0; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.cloud.servicebroker.model.catalog.Catalog' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:769) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:509) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:849) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877) ~[spring-context-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549) ~[spring-context-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:127) ~[spring-boot-test-2.1.4.RELEASE.jar:2.1.4.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99) ~[spring-test-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117) ~[spring-test-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	... 28 common frames omitted
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.cloud.servicebroker.model.catalog.Catalog' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoMatchingBeanFound(DefaultListableBeanFactory.java:1654) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1213) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1167) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
	... 46 common frames omitted
```

Rather than failing, the auto-configuration should probably back off as it used to in `2.0.x`.

Humorously, that was a request to fail faster with bad configuration. 

Sent from my phone

> On May 4, 2019, at 07:29, Stéphane Nicoll <notifications@github.com> wrote:
> 
> If I generate an app from start.spring.io with version 2.1.2.RELEASE, I can build it and run it. However if I upgrade to 3.0.0.RELEASE I get the following:
> 
> Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'beanCatalogService' defined in class path resource [org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfiguration.class]: Unsatisfied dependency expressed through method 'beanCatalogService' parameter 0; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.cloud.servicebroker.model.catalog.Catalog' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
> 	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:769) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:509) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:849) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877) ~[spring-context-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549) ~[spring-context-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
> 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
> 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316) ~[spring-boot-2.1.4.RELEASE.jar:2.1.4.RELEASE]
> 	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:127) ~[spring-boot-test-2.1.4.RELEASE.jar:2.1.4.RELEASE]
> 	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99) ~[spring-test-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117) ~[spring-test-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	... 28 common frames omitted
> Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.cloud.servicebroker.model.catalog.Catalog' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoMatchingBeanFound(DefaultListableBeanFactory.java:1654) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1213) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1167) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760) ~[spring-beans-5.1.6.RELEASE.jar:5.1.6.RELEASE]
> 	... 46 common frames omitted
> Rather than failing, the auto-configuration should probably back off as it used to in 2.0.x.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/RequiredCatalogBeanFailureAnalyzer.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/RequiredCatalogBeanFailureAnalyzer.java
@@ -1,0 +1,27 @@
+package org.springframework.cloud.servicebroker.autoconfigure.web;
+
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.CatalogDefinitionDoesNotExistException;
+import org.springframework.cloud.servicebroker.model.catalog.Catalog;
+import org.springframework.cloud.servicebroker.service.CatalogService;
+
+/**
+ * @author Roy Clarkson
+ */
+public class RequiredCatalogBeanFailureAnalyzer extends AbstractFailureAnalyzer<CatalogDefinitionDoesNotExistException> {
+
+	private static final String REFERENCE_DOC = "https://docs.spring.io/spring-cloud-open-service-broker/docs/current" +
+			"/reference/html5/#service-catalog";
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, CatalogDefinitionDoesNotExistException cause) {
+		String description = "A 'service broker catalog' is required for Spring Cloud Open Service Broker applications";
+		String action = String.format("Consider defining a catalog in properties or a bean of type" +
+						" '%s' in your configuration. Alternatively, you may implement a service of type '%s'. See " +
+						"the reference documentation for more information: " + REFERENCE_DOC, Catalog.class,
+				CatalogService.class);
+		return new FailureAnalysis(description, action, cause);
+	}
+
+}

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/RequiredServiceInstanceServiceBeanFailureAnalyzer.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/RequiredServiceInstanceServiceBeanFailureAnalyzer.java
@@ -1,0 +1,30 @@
+package org.springframework.cloud.servicebroker.autoconfigure.web;
+
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.CatalogDefinitionDoesNotExistException;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.ServiceInstanceServiceBeanDoesNotExistException;
+import org.springframework.cloud.servicebroker.model.catalog.Catalog;
+import org.springframework.cloud.servicebroker.service.CatalogService;
+import org.springframework.cloud.servicebroker.service.ServiceInstanceService;
+
+/**
+ * @author Roy Clarkson
+ */
+public class RequiredServiceInstanceServiceBeanFailureAnalyzer
+		extends AbstractFailureAnalyzer<ServiceInstanceServiceBeanDoesNotExistException> {
+
+	private static final String REFERENCE_DOC = "https://docs.spring.io/spring-cloud-open-service-broker/docs/current" +
+			"/reference/html5/#service-instances";
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, ServiceInstanceServiceBeanDoesNotExistException cause) {
+		String description = String.format("Service brokers must implement the '%s' and " +
+				"provide implementations of the required methods of that interface.", ServiceInstanceService.class);
+		String action = String.format("Consider defining a bean of type '%s' in your configuration. See " +
+						"the reference documentation for more information: " + REFERENCE_DOC,
+				ServiceInstanceService.class);
+		return new FailureAnalysis(description, action, cause);
+	}
+
+}

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfiguration.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfiguration.java
@@ -16,10 +16,12 @@
 
 package org.springframework.cloud.servicebroker.autoconfigure.web;
 
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.CatalogDefinitionDoesNotExistException;
 import org.springframework.cloud.servicebroker.model.catalog.Catalog;
 import org.springframework.cloud.servicebroker.service.BeanCatalogService;
 import org.springframework.cloud.servicebroker.service.CatalogService;
@@ -48,7 +50,7 @@ import org.springframework.context.annotation.Configuration;
 public class ServiceBrokerAutoConfiguration {
 
 	@Configuration
-	@ConditionalOnMissingBean({ Catalog.class, CatalogService.class })
+	@ConditionalOnMissingBean({Catalog.class, CatalogService.class})
 	@EnableConfigurationProperties(ServiceBrokerProperties.class)
 	@ConditionalOnProperty(prefix = "spring.cloud.openservicebroker.catalog.services[0]", name = "id")
 	protected static class CatalogPropertiesMinimalConfiguration {
@@ -67,7 +69,10 @@ public class ServiceBrokerAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean(CatalogService.class)
-	public CatalogService beanCatalogService(Catalog catalog) {
+	public CatalogService beanCatalogService(@Autowired(required = false) Catalog catalog) {
+		if (catalog == null) {
+			throw new CatalogDefinitionDoesNotExistException();
+		}
 		return new BeanCatalogService(catalog);
 	}
 

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/exception/CatalogDefinitionDoesNotExistException.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/exception/CatalogDefinitionDoesNotExistException.java
@@ -1,0 +1,9 @@
+package org.springframework.cloud.servicebroker.autoconfigure.web.exception;
+
+/**
+ * @author Roy Clarkson
+ */
+public class CatalogDefinitionDoesNotExistException extends RuntimeException {
+
+	private static final long serialVersionUID = -6345978650178340540L;
+}

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/exception/ServiceInstanceServiceBeanDoesNotExistException.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/exception/ServiceInstanceServiceBeanDoesNotExistException.java
@@ -1,0 +1,9 @@
+package org.springframework.cloud.servicebroker.autoconfigure.web.exception;
+
+/**
+ * @author Roy Clarkson
+ */
+public class ServiceInstanceServiceBeanDoesNotExistException extends RuntimeException {
+
+	private static final long serialVersionUID = -5517038451033555513L;
+}

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/reactive/ServiceBrokerWebFluxAutoConfiguration.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/reactive/ServiceBrokerWebFluxAutoConfiguration.java
@@ -16,12 +16,14 @@
 
 package org.springframework.cloud.servicebroker.autoconfigure.web.reactive;
 
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.AutoConfigureAfter;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
 import org.springframework.boot.autoconfigure.web.reactive.WebFluxAutoConfiguration;
 import org.springframework.cloud.servicebroker.autoconfigure.web.ServiceBrokerAutoConfiguration;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.ServiceInstanceServiceBeanDoesNotExistException;
 import org.springframework.cloud.servicebroker.controller.CatalogController;
 import org.springframework.cloud.servicebroker.controller.ServiceBrokerExceptionHandler;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceBindingController;
@@ -54,9 +56,12 @@ public class ServiceBrokerWebFluxAutoConfiguration {
 	private final ServiceInstanceBindingEventService serviceInstanceBindingEventService;
 
 	protected ServiceBrokerWebFluxAutoConfiguration(CatalogService catalogService,
-			ServiceInstanceService serviceInstanceService,
+			@Autowired(required = false) ServiceInstanceService serviceInstanceService,
 			ServiceInstanceBindingService serviceInstanceBindingService,
 			EventFlowRegistries eventFlowRegistries) {
+		if (serviceInstanceService == null) {
+			throw new ServiceInstanceServiceBeanDoesNotExistException();
+		}
 		this.catalogService = catalogService;
 		this.serviceInstanceEventService = new ServiceInstanceEventService(
 				serviceInstanceService, eventFlowRegistries);

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/servlet/ServiceBrokerWebMvcAutoConfiguration.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/java/org/springframework/cloud/servicebroker/autoconfigure/web/servlet/ServiceBrokerWebMvcAutoConfiguration.java
@@ -16,12 +16,14 @@
 
 package org.springframework.cloud.servicebroker.autoconfigure.web.servlet;
 
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.AutoConfigureAfter;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
 import org.springframework.boot.autoconfigure.web.servlet.WebMvcAutoConfiguration;
 import org.springframework.cloud.servicebroker.autoconfigure.web.ServiceBrokerAutoConfiguration;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.ServiceInstanceServiceBeanDoesNotExistException;
 import org.springframework.cloud.servicebroker.controller.CatalogController;
 import org.springframework.cloud.servicebroker.controller.ServiceBrokerExceptionHandler;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceBindingController;
@@ -55,9 +57,12 @@ public class ServiceBrokerWebMvcAutoConfiguration {
 	private final ServiceInstanceBindingEventService serviceInstanceBindingEventService;
 
 	protected ServiceBrokerWebMvcAutoConfiguration(CatalogService catalogService,
-			ServiceInstanceService serviceInstanceService,
+			@Autowired(required = false) ServiceInstanceService serviceInstanceService,
 			ServiceInstanceBindingService serviceInstanceBindingService,
 			EventFlowRegistries eventFlowRegistries) {
+		if (serviceInstanceService == null) {
+			throw new ServiceInstanceServiceBeanDoesNotExistException();
+		}
 		this.catalogService = catalogService;
 		this.serviceInstanceEventService = new ServiceInstanceEventService(
 				serviceInstanceService, eventFlowRegistries);

--- a/spring-cloud-open-service-broker-autoconfigure/src/main/resources/META-INF/spring.factories
+++ b/spring-cloud-open-service-broker-autoconfigure/src/main/resources/META-INF/spring.factories
@@ -4,3 +4,7 @@ org.springframework.cloud.servicebroker.autoconfigure.web.servlet.ServiceBrokerW
 org.springframework.cloud.servicebroker.autoconfigure.web.servlet.ApiVersionWebMvcAutoConfiguration,\
 org.springframework.cloud.servicebroker.autoconfigure.web.reactive.ServiceBrokerWebFluxAutoConfiguration,\
 org.springframework.cloud.servicebroker.autoconfigure.web.reactive.ApiVersionWebFluxAutoConfiguration
+
+org.springframework.boot.diagnostics.FailureAnalyzer=\
+org.springframework.cloud.servicebroker.autoconfigure.web.RequiredCatalogBeanFailureAnalyzer,\
+org.springframework.cloud.servicebroker.autoconfigure.web.RequiredServiceInstanceServiceBeanFailureAnalyzer

--- a/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/AbstractServiceBrokerWebAutoConfigurationTest.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/AbstractServiceBrokerWebAutoConfigurationTest.java
@@ -1,0 +1,112 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.servicebroker.autoconfigure.web;
+
+import org.junit.Test;
+
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.cloud.servicebroker.autoconfigure.web.reactive.ServiceBrokerWebFluxAutoConfiguration;
+import org.springframework.cloud.servicebroker.autoconfigure.web.servlet.ServiceBrokerWebMvcAutoConfiguration;
+import org.springframework.cloud.servicebroker.controller.CatalogController;
+import org.springframework.cloud.servicebroker.controller.ServiceInstanceBindingController;
+import org.springframework.cloud.servicebroker.controller.ServiceInstanceController;
+import org.springframework.cloud.servicebroker.service.CatalogService;
+import org.springframework.cloud.servicebroker.service.ServiceInstanceBindingService;
+import org.springframework.cloud.servicebroker.service.ServiceInstanceService;
+import org.springframework.cloud.servicebroker.service.events.EventFlowRegistries;
+import org.springframework.context.annotation.Bean;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public abstract class AbstractServiceBrokerWebAutoConfigurationTest {
+
+	private static final String ANALYZER_DESCRIPTION = String.format("Service brokers must implement the '%s' and " +
+			"provide implementations of the required methods of that interface.", ServiceInstanceService.class);
+
+	@Test
+	public void controllersAreNotCreatedWithNonWebConfiguration() {
+		nonWebApplicationContextRunner()
+				.withUserConfiguration(FullServicesConfiguration.class)
+				.run(context -> assertThat(context).doesNotHaveBean(CatalogController.class)
+						.doesNotHaveBean(ServiceInstanceController.class)
+						.doesNotHaveBean(ServiceInstanceBindingController.class));
+	}
+
+	protected void assertFailureAnalysis(Throwable t) {
+		FailureAnalyzer analyzer = new RequiredServiceInstanceServiceBeanFailureAnalyzer();
+		FailureAnalysis analysis = analyzer.analyze(t);
+		assertThat(analysis).isNotNull();
+		assertThat(analysis.getDescription()).isEqualTo(ANALYZER_DESCRIPTION);
+	}
+
+	protected AutoConfigurations autoConfigurations() {
+		return AutoConfigurations.of(ServiceBrokerWebFluxAutoConfiguration.class,
+				ServiceBrokerWebMvcAutoConfiguration.class);
+	}
+
+	protected ApplicationContextRunner nonWebApplicationContextRunner() {
+		return new ApplicationContextRunner().withConfiguration(autoConfigurations());
+	}
+
+	@TestConfiguration
+	public static class FullServicesConfiguration {
+
+		@Bean
+		public CatalogService catalogService() {
+			return new TestCatalogService();
+		}
+
+		@Bean
+		public ServiceInstanceService serviceInstanceService() {
+			return new TestServiceInstanceService();
+		}
+
+		@Bean
+		public ServiceInstanceBindingService serviceInstanceBindingService() {
+			return new TestServiceInstanceBindingService();
+		}
+
+		@Bean
+		public EventFlowRegistries eventFlowRegistries() {
+			return new EventFlowRegistries();
+		}
+	}
+
+	@TestConfiguration
+	public static class MissingServiceInstanceServiceConfiguration {
+
+		@Bean
+		public CatalogService catalogService() {
+			return new TestCatalogService();
+		}
+
+		@Bean
+		public ServiceInstanceBindingService serviceInstanceBindingService() {
+			return new TestServiceInstanceBindingService();
+		}
+
+		@Bean
+		public EventFlowRegistries eventFlowRegistries() {
+			return new EventFlowRegistries();
+		}
+	}
+
+}

--- a/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfigurationTest.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/ServiceBrokerAutoConfigurationTest.java
@@ -20,11 +20,13 @@ import static org.assertj.core.api.Assertions.assertThat;
 
 import org.junit.Test;
 
-import org.springframework.beans.factory.BeanCreationException;
 import org.springframework.beans.factory.UnsatisfiedDependencyException;
 import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
 import org.springframework.boot.test.context.TestConfiguration;
 import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.CatalogDefinitionDoesNotExistException;
 import org.springframework.cloud.servicebroker.model.catalog.Catalog;
 import org.springframework.cloud.servicebroker.service.BeanCatalogService;
 import org.springframework.cloud.servicebroker.service.CatalogService;
@@ -35,6 +37,9 @@ import org.springframework.cloud.servicebroker.service.events.EventFlowRegistrie
 import org.springframework.context.annotation.Bean;
 
 public class ServiceBrokerAutoConfigurationTest {
+
+	private static final String ANALYZER_DESCRIPTION = "A 'service broker catalog' is required for Spring Cloud Open" +
+			" Service Broker applications";
 
 	private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
 			.withConfiguration(AutoConfigurations.of(ServiceBrokerAutoConfiguration.class));
@@ -138,9 +143,20 @@ public class ServiceBrokerAutoConfigurationTest {
 	@Test
 	public void servicesAreNotCreatedWhenMissingCatalogAndCatalogServiceConfiguration() {
 		this.contextRunner
-				.withUserConfiguration(MissingCatalogAndCatalogServiceConfiguration.class)
-				.run((context) -> assertThat(context.getStartupFailure())
-						.isExactlyInstanceOf(UnsatisfiedDependencyException.class));
+				.withUserConfiguration(MissingCatalogServiceConfiguration.class)
+				.run((context) -> {
+					Throwable t = context.getStartupFailure();
+					assertThat(t).isExactlyInstanceOf(UnsatisfiedDependencyException.class)
+							.hasRootCauseExactlyInstanceOf(CatalogDefinitionDoesNotExistException.class);
+					assertFailureAnalysis(t);
+				});
+	}
+
+	private void assertFailureAnalysis(Throwable t) {
+		FailureAnalyzer analyzer = new RequiredCatalogBeanFailureAnalyzer();
+		FailureAnalysis analysis = analyzer.analyze(t);
+		assertThat(analysis).isNotNull();
+		assertThat(analysis.getDescription()).isEqualTo(ANALYZER_DESCRIPTION);
 	}
 
 	@Test
@@ -154,7 +170,7 @@ public class ServiceBrokerAutoConfigurationTest {
 	@Test
 	public void servicesAreCreatedFromCatalogProperties() {
 		this.contextRunner
-				.withUserConfiguration(MissingCatalogAndCatalogServiceConfiguration.class)
+				.withUserConfiguration(MissingCatalogServiceConfiguration.class)
 				.withPropertyValues(
 						"spring.cloud.openservicebroker.catalog.services[0].id=service-one-id",
 						"spring.cloud.openservicebroker.catalog.services[0].name=Service One",
@@ -268,8 +284,8 @@ public class ServiceBrokerAutoConfigurationTest {
 	}
 
 	@TestConfiguration
-	public static class MissingCatalogAndCatalogServiceConfiguration {
-		public MissingCatalogAndCatalogServiceConfiguration(Catalog catalog, CatalogService catalogService) {
+	public static class MissingCatalogServiceConfiguration {
+		public MissingCatalogServiceConfiguration(CatalogService catalogService) {
 		}
 		@Bean
 		public ServiceInstanceService serviceInstanceService() {
@@ -279,7 +295,7 @@ public class ServiceBrokerAutoConfigurationTest {
 
 	@TestConfiguration
 	public static class MissingAllConfiguration {
-		public MissingAllConfiguration(Catalog catalog, CatalogService catalogService, ServiceInstanceService serviceInstanceService) {
+		public MissingAllConfiguration(CatalogService catalogService, ServiceInstanceService serviceInstanceService) {
 		}
 	}
 

--- a/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/reactive/ServiceBrokerWebFluxAutoConfigurationTest.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/reactive/ServiceBrokerWebFluxAutoConfigurationTest.java
@@ -18,90 +18,53 @@ package org.springframework.cloud.servicebroker.autoconfigure.web.reactive;
 
 import org.junit.Test;
 
+import org.springframework.beans.factory.BeanCreationException;
 import org.springframework.beans.factory.UnsatisfiedDependencyException;
-import org.springframework.boot.autoconfigure.AutoConfigurations;
-import org.springframework.boot.test.context.TestConfiguration;
-import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
 import org.springframework.boot.test.context.runner.ReactiveWebApplicationContextRunner;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestCatalogService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestServiceInstanceBindingService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestServiceInstanceService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.servlet.ServiceBrokerWebMvcAutoConfiguration;
+import org.springframework.cloud.servicebroker.autoconfigure.web.AbstractServiceBrokerWebAutoConfigurationTest;
+import org.springframework.cloud.servicebroker.autoconfigure.web.RequiredServiceInstanceServiceBeanFailureAnalyzer;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.ServiceInstanceServiceBeanDoesNotExistException;
 import org.springframework.cloud.servicebroker.controller.CatalogController;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceBindingController;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceController;
-import org.springframework.cloud.servicebroker.service.CatalogService;
-import org.springframework.cloud.servicebroker.service.ServiceInstanceBindingService;
 import org.springframework.cloud.servicebroker.service.ServiceInstanceService;
-import org.springframework.cloud.servicebroker.service.events.EventFlowRegistries;
-import org.springframework.context.annotation.Bean;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-public class ServiceBrokerWebFluxAutoConfigurationTest {
-
-	@Test
-	public void controllersAreNotCreatedWithNonWebConfiguration() {
-		nonWebApplicationContextRunner()
-				.withUserConfiguration(FullServicesConfiguration.class)
-				.run((context) -> {
-					assertThat(context).doesNotHaveBean(CatalogController.class);
-					assertThat(context).doesNotHaveBean(ServiceInstanceController.class);
-					assertThat(context).doesNotHaveBean(ServiceInstanceBindingController.class);
-				});
-	}
+public class ServiceBrokerWebFluxAutoConfigurationTest extends AbstractServiceBrokerWebAutoConfigurationTest {
 
 	@Test
 	public void controllersAreNotCreatedWithoutRequiredServices() {
 		webApplicationContextRunner()
 				.run(context -> assertThat(context.getStartupFailure())
-							.isExactlyInstanceOf(UnsatisfiedDependencyException.class));
+						.isExactlyInstanceOf(UnsatisfiedDependencyException.class));
 	}
 
 	@Test
 	public void controllersAreCreated() {
 		webApplicationContextRunner()
 				.withUserConfiguration(FullServicesConfiguration.class)
-				.run((context) -> {
-					assertThat(context).hasSingleBean(CatalogController.class);
-					assertThat(context).hasSingleBean(ServiceInstanceController.class);
-					assertThat(context).hasSingleBean(ServiceInstanceBindingController.class);
+				.run(context -> assertThat(context).hasSingleBean(CatalogController.class)
+						.hasSingleBean(ServiceInstanceController.class)
+						.hasSingleBean(ServiceInstanceBindingController.class));
+	}
+
+	@Test
+	public void controllersAreNotCreatedWithMissingInstanceService() {
+		webApplicationContextRunner()
+				.withUserConfiguration(MissingServiceInstanceServiceConfiguration.class)
+				.run(context -> {
+					Throwable t = context.getStartupFailure();
+					assertThat(t).isExactlyInstanceOf(BeanCreationException.class)
+							.hasRootCauseExactlyInstanceOf(ServiceInstanceServiceBeanDoesNotExistException.class);
+					assertFailureAnalysis(t);
 				});
 	}
 
 	private ReactiveWebApplicationContextRunner webApplicationContextRunner() {
-		return new ReactiveWebApplicationContextRunner().withConfiguration(
-				AutoConfigurations.of(ServiceBrokerWebFluxAutoConfiguration.class,
-						ServiceBrokerWebMvcAutoConfiguration.class));
-	}
-
-	private ApplicationContextRunner nonWebApplicationContextRunner() {
-		return new ApplicationContextRunner().withConfiguration(
-				AutoConfigurations.of(ServiceBrokerWebFluxAutoConfiguration.class,
-						ServiceBrokerWebMvcAutoConfiguration.class));
-	}
-
-	@TestConfiguration
-	public static class FullServicesConfiguration {
-		@Bean
-		public CatalogService catalogService() {
-			return new TestCatalogService();
-		}
-
-		@Bean
-		public ServiceInstanceService serviceInstanceService() {
-			return new TestServiceInstanceService();
-		}
-
-		@Bean
-		public ServiceInstanceBindingService serviceInstanceBindingService() {
-			return new TestServiceInstanceBindingService();
-		}
-
-		@Bean
-		public EventFlowRegistries eventFlowRegistries() {
-			return new EventFlowRegistries();
-		}
+		return new ReactiveWebApplicationContextRunner().withConfiguration(autoConfigurations());
 	}
 
 }

--- a/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/servlet/ServiceBrokerWebMvcAutoConfigurationTest.java
+++ b/spring-cloud-open-service-broker-autoconfigure/src/test/java/org/springframework/cloud/servicebroker/autoconfigure/web/servlet/ServiceBrokerWebMvcAutoConfigurationTest.java
@@ -18,89 +18,49 @@ package org.springframework.cloud.servicebroker.autoconfigure.web.servlet;
 
 import org.junit.Test;
 
+import org.springframework.beans.factory.BeanCreationException;
 import org.springframework.beans.factory.UnsatisfiedDependencyException;
-import org.springframework.boot.autoconfigure.AutoConfigurations;
-import org.springframework.boot.test.context.TestConfiguration;
-import org.springframework.boot.test.context.runner.ApplicationContextRunner;
 import org.springframework.boot.test.context.runner.WebApplicationContextRunner;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestCatalogService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestServiceInstanceBindingService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.TestServiceInstanceService;
-import org.springframework.cloud.servicebroker.autoconfigure.web.reactive.ServiceBrokerWebFluxAutoConfiguration;
+import org.springframework.cloud.servicebroker.autoconfigure.web.AbstractServiceBrokerWebAutoConfigurationTest;
+import org.springframework.cloud.servicebroker.autoconfigure.web.exception.ServiceInstanceServiceBeanDoesNotExistException;
 import org.springframework.cloud.servicebroker.controller.CatalogController;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceBindingController;
 import org.springframework.cloud.servicebroker.controller.ServiceInstanceController;
-import org.springframework.cloud.servicebroker.service.CatalogService;
-import org.springframework.cloud.servicebroker.service.ServiceInstanceBindingService;
-import org.springframework.cloud.servicebroker.service.ServiceInstanceService;
-import org.springframework.cloud.servicebroker.service.events.EventFlowRegistries;
-import org.springframework.context.annotation.Bean;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-public class ServiceBrokerWebMvcAutoConfigurationTest {
-
-	@Test
-	public void controllersAreNotCreatedWithNonWebConfiguration() {
-		nonWebApplicationContextRunner()
-				.withUserConfiguration(FullServicesConfiguration.class)
-				.run((context) -> {
-					assertThat(context).doesNotHaveBean(CatalogController.class);
-					assertThat(context).doesNotHaveBean(ServiceInstanceController.class);
-					assertThat(context).doesNotHaveBean(ServiceInstanceBindingController.class);
-				});
-	}
+public class ServiceBrokerWebMvcAutoConfigurationTest extends AbstractServiceBrokerWebAutoConfigurationTest {
 
 	@Test
 	public void controllersAreNotCreatedWithoutRequiredServices() {
 		webApplicationContextRunner()
 				.run(context -> assertThat(context.getStartupFailure())
-						.isInstanceOf(UnsatisfiedDependencyException.class));
+						.isExactlyInstanceOf(UnsatisfiedDependencyException.class));
 	}
 
 	@Test
 	public void controllersAreCreated() {
 		webApplicationContextRunner()
 				.withUserConfiguration(FullServicesConfiguration.class)
-				.run((context) -> {
-					assertThat(context).hasSingleBean(CatalogController.class);
-					assertThat(context).hasSingleBean(ServiceInstanceController.class);
-					assertThat(context).hasSingleBean(ServiceInstanceBindingController.class);
+				.run(context -> assertThat(context).hasSingleBean(CatalogController.class)
+						.hasSingleBean(ServiceInstanceController.class)
+						.hasSingleBean(ServiceInstanceBindingController.class));
+	}
+
+	@Test
+	public void controllersAreNotCreatedWithMissingInstanceService() {
+		webApplicationContextRunner()
+				.withUserConfiguration(MissingServiceInstanceServiceConfiguration.class)
+				.run(context -> {
+					Throwable t = context.getStartupFailure();
+					assertThat(t).isExactlyInstanceOf(BeanCreationException.class)
+							.hasRootCauseExactlyInstanceOf(ServiceInstanceServiceBeanDoesNotExistException.class);
+					assertFailureAnalysis(t);
 				});
 	}
 
 	private WebApplicationContextRunner webApplicationContextRunner() {
-		return new WebApplicationContextRunner().withConfiguration(
-				AutoConfigurations.of(ServiceBrokerWebMvcAutoConfiguration.class,
-						ServiceBrokerWebFluxAutoConfiguration.class));
+		return new WebApplicationContextRunner().withConfiguration(autoConfigurations());
 	}
 
-	private ApplicationContextRunner nonWebApplicationContextRunner() {
-		return new ApplicationContextRunner().withConfiguration(
-				AutoConfigurations.of(ServiceBrokerWebMvcAutoConfiguration.class,
-						ServiceBrokerWebFluxAutoConfiguration.class));
-	}
-
-	@TestConfiguration
-	public static class FullServicesConfiguration {
-		@Bean
-		public CatalogService catalogService() {
-			return new TestCatalogService();
-		}
-
-		@Bean
-		public ServiceInstanceService serviceInstanceService() {
-			return new TestServiceInstanceService();
-		}
-
-		@Bean
-		public ServiceInstanceBindingService serviceInstanceBindingService() {
-			return new TestServiceInstanceBindingService();
-		}
-
-		@Bean
-		public EventFlowRegistries eventFlowRegistries() {
-			return new EventFlowRegistries();
-		}
-	}
 }

